# Two-Qubit Hadamard Superposition and Statevector Analysis

## 1. Problem Statement
Create a 2-qubit quantum circuit initialized in $|00\rangle$, apply a Hadamard gate to both qubits, obtain the resulting statevector, display the amplitudes of all computational basis states, and verify that the statevector is normalized.

---

## 2. Theoretical Background & Concepts

### Initial 2-Qubit State
In quantum computing and Qiskit, multi-qubit systems are initialized by default in the all-zero computational ground state:

$$|00\rangle = |0\rangle \otimes |0\rangle = \begin{pmatrix} 1 \\ 0 \\ 0 \\ 0 \end{pmatrix}$$

### Creating an Equal 2-Qubit Superposition
The **Hadamard gate** ($H$) creates an equal superposition from computational basis states. When applied independently to both qubits of a system initialized in $|00\rangle$, the transformation corresponds to the tensor product operator $H \otimes H = H^{\otimes 2}$:

$$|\psi\rangle = (H \otimes H)|00\rangle = (H|0\rangle) \otimes (H|0\rangle)$$

Expanding each single-qubit state $H|0\rangle = \frac{|0\rangle + |1\rangle}{\sqrt{2}}$:

$$|\psi\rangle = \left(\frac{|0\rangle + |1\rangle}{\sqrt{2}}\right) \otimes \left(\frac{|0\rangle + |1\rangle}{\sqrt{2}}\right) = \frac{1}{2}(|00\rangle + |01\rangle + |10\rangle + |11\rangle)$$

Applying Hadamard gates to both qubits produces an equal linear superposition across all $2^2 = 4$ computational basis states.

### Amplitudes and Measurement Probabilities
A **statevector** contains the complex probability amplitudes $c_x$ associated with the computational basis states:

$$|\psi\rangle = c_{00}|00\rangle + c_{01}|01\rangle + c_{10}|10\rangle + c_{11}|11\rangle = \begin{pmatrix} c_{00} \\ c_{01} \\ c_{10} \\ c_{11} \end{pmatrix}$$

For this equal superposition state:
- Each computational basis state has an equal amplitude:
  $$c_{00} = c_{01} = c_{10} = c_{11} = \frac{1}{2} = 0.5$$
- According to the **Born rule**, the probability of measuring each basis state upon wavefunction collapse is the squared magnitude of its amplitude:
  $$P(x) = |c_x|^2 = \left|\frac{1}{2}\right|^2 = \frac{1}{4} = 0.25 \quad (25\%)$$
  for each state $x \in \{00, 01, 10, 11\}$.

### Statevector Normalization
A foundational postulate of quantum mechanics requires that the total probability of all possible measurement outcomes must equal 1:

$$\sum_{x \in \{00, 01, 10, 11\}} P(x) = \sum_{x} |c_x|^2 = |c_{00}|^2 + |c_{01}|^2 + |c_{10}|^2 + |c_{11}|^2 = 1$$

Verifying that $\sum |c_x|^2 = 1$ confirms that the statevector is properly normalized and represents a physically valid quantum state.

---

## 3. Circuit Flow
The transformation sequence is:

$$|00\rangle \xrightarrow{\quad H \text{ on qubit 0 and } H \text{ on qubit 1} \quad} |\psi\rangle = \frac{1}{2}(|00\rangle + |01\rangle + |10\rangle + |11\rangle) \quad \text{(equal 2-qubit superposition)}$$

---

## 4. Tools & Libraries Used
- **Qiskit**: IBM's open-source quantum computing framework for constructing and visualizing quantum circuits (`QuantumCircuit`).
- **`qiskit.quantum_info.Statevector`**: Analytical module for extracting exact quantum statevectors directly from circuit instructions without measurement collapse.
- **NumPy**: Numerical computation library used for vector operations and verifying normalization ($\sum |c_x|^2$).

### Step 1: Import Required Libraries
Import `QuantumCircuit` and `Statevector` from Qiskit, and `numpy` for numerical analysis.

In [1]:
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
import numpy as np


### Step 2: Create the 2-Qubit Quantum Circuit
Initialize a 2-qubit quantum circuit. By convention, both qubits start in the default ground state $|00\rangle$.

In [2]:
# Create circuit
qc = QuantumCircuit(2)


### Step 3: Apply Hadamard Gates
Apply a Hadamard gate to qubit 0 (`qc.h(0)`) and qubit 1 (`qc.h(1)`) to prepare an equal superposition across both qubits.

In [3]:
# Apply Hadamard to both qubits
qc.h(0)
qc.h(1)


### Step 4: Obtain the Statevector
Extract the exact quantum statevector from the circuit instructions using `Statevector.from_instruction(qc)`, and display the circuit diagram and statevector.

In [4]:
# Get statevector
state = Statevector.from_instruction(qc)

print("Quantum Circuit:")
print(qc)

print("\nStatevector:")
print(state)


Quantum Circuit:
     ┌───┐
q_0: ┤ H ├
     ├───┤
q_1: ┤ H ├
     └───┘

Statevector:
Statevector([0.5+0.j, 0.5+0.j, 0.5+0.j, 0.5+0.j],
            dims=(2, 2))


### Step 5: Display Basis-State Amplitudes
Iterate over the statevector amplitudes (`state.data`) and format each basis state with its 2-bit binary label (`|00>`, `|01>`, `|10>`, `|11>`).

In [5]:
# Display amplitudes
print("\nAmplitudes:")
for i, amp in enumerate(state.data):
    print(f"|{i:02b}> : {amp}")



Amplitudes:
|00> : (0.4999999999999999+0j)
|01> : (0.4999999999999999+0j)
|10> : (0.4999999999999999+0j)
|11> : (0.4999999999999999+0j)


### Step 6: Verify Statevector Normalization
Calculate the sum of the squared absolute magnitudes of all state amplitudes using `np.sum(np.abs(state.data)**2)` to verify probability conservation.

In [6]:
# Verify normalization
probability = np.sum(np.abs(state.data)**2)

print("\nSum of probabilities =", probability)



Sum of probabilities = 0.9999999999999996


## Conclusion

- **Equal 2-Qubit Superposition**: Applying Hadamard gates to both qubits transforms the initial ground state $|00\rangle$ into an equal superposition across all four computational basis states:
  $$|\psi\rangle = \frac{1}{2}(|00\rangle + |01\rangle + |10\rangle + |11\rangle)$$
- **Uniform Amplitudes & Probabilities**: Each computational basis state ($|00\rangle, |01\rangle, |10\rangle, |11\rangle$) has an identical probability amplitude of $c \approx 0.5$, resulting in an equal theoretical probability of $P = |c|^2 = 0.25$ ($25\%$) per state.
- **Normalization Confirmation**: The sum of the squared magnitudes evaluates to $\sum |c_x|^2 = 0.9999999999999996 \approx 1.0$, mathematically confirming that total probability is conserved and that the statevector represents a physically valid quantum state.